Cython/Utility/ObjectHandling.c

```diff
@@ -1024,6 +1024,91 @@ static PyObject *__Pyx_CreateClass(PyObject *bases, PyObject *dict, PyObject *na
     return result;
 }
 
+/////////////// Py3UpdateBases.proto ///////////////
+
+static PyObject* __Pyx_PEP560_update_bases(PyObject *bases); /* proto */
+
+/////////////// Py3UpdateBases /////////////////////
+//@requires: PyObjectCallOneArg
+//@requires: PyObjectGetAttrStrNoError
+
+/* Shamelessly adapted from cpython/bltinmodule.c update_bases */
+static PyObject*
+__Pyx_PEP560_update_bases(PyObject *bases)
+{
+    Py_ssize_t i, j;
+    PyObject *base, *meth, *new_base, *result, *new_bases = NULL;
+    /*assert(PyTuple_Check(bases));*/
+
+    for (i = 0; i < PyTuple_GET_SIZE(bases); i++) {
+        base  = PyTuple_GET_ITEM(bases, i); /* original code in CPython: base  = args[i]; */
+        if (PyType_Check(base)) {
+            if (new_bases) {
+                /* If we already have made a replacement, then we append every normal base,
+                   otherwise just skip it. */
+                if (PyList_Append(new_bases, base) < 0) {
+                    goto error;
+                }
+            }
+            continue;
+        }
+        /* original code in CPython:
+         * if (_PyObject_LookupAttrId(base, &PyId___mro_entries__, &meth) < 0) {
+         */
+        meth = __Pyx_PyObject_GetAttrStrNoError(base, PYIDENT("__mro_entries__"));
```

Comment:
https://github.com/python/cpython/blob/32bd68c839adb7b42af12366ab0892303115d1d1/Objects/object.c#L916 `_PyObject_LookupAttr` does look to call `getattr` and `gettattro` if they exist.

I'm sure it's not a perfect replacement though. That's mainly why I left the "original code in CPython" comments, so we'd stand a better chance of working out what's gone wrong when one of the replacements isn't right.

Reply:
Ah, ok, I stand corrected. I was confusing it with `PyType_Lookup()`. I guess the `Lookup` hint just tries to set the name apart from the usual signature that returns the object directly, instead of an error marker as done here.